server/app/services/docker/streaming_executor.rb

```diff
@@ -0,0 +1,70 @@
+module Docker
+  class StreamingExecutor
+
+    # @param [Container] container
+    # @param [#send,#close] ws
+    def initialize(container, ws)
+      @container = container
+      @ws = ws
+      @client = RpcClient.new(container.host_node.node_id)
+    end
+
+    # Starts exec session with tty
+    def start_tty
+      create_session
+      subscribe_to_session
+      register_tty_ws_events
+    end
+
+    # Starts normal exec session (without tty)
+    def start 
+      create_session
+      subscribe_to_session
+      register_run_ws_events
+    end
+
+    def create_session
+      @exec_session = @client.request('/containers/create_exec', @container.container_id)
+    end
+
+    def subscribe_to_session
+      @subscription = MongoPubsub.subscribe("container_exec:#{@exec_session['id']}") do |data|
+        if data.has_key?('exit')
+          @ws.send(JSON.dump({ exit: data['exit'] }))
+          @ws.close
+        else
+          @ws.send(JSON.dump({ stream: data['stream'], chunk: data['chunk'] }))
+        end
+      end
+    end
+
+    def register_tty_ws_events
+      init = true
+      @ws.on(:message) do |event|
+        if init == true
+          cmd = ['/bin/sh', '-c', event.data]
```

Comment:
I take this back, the websocket connection must be a GET request and can't have any POST request body parameters.

It might still be possible to take the command as query parameter or header, but reading the command from the first websocket frame also makes sense.